src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Rename/RenamerProjectTreeActionHandler.cs

```diff
@@ -0,0 +1,203 @@
+
+using System.ComponentModel.Composition;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading.Tasks;
+using EnvDTE;
+using Microsoft.CodeAnalysis;
+using Microsoft.VisualStudio.LanguageServices;
+using Microsoft.VisualStudio.ProjectSystem.Rename;
+using Microsoft.VisualStudio.ProjectSystem.Waiting;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.Rename
+{
+    [Order(Order.Default)]
+    [Export(typeof(IProjectTreeActionHandler))]
+    [Export(typeof(IFileRenameHandler))]
+    [AppliesTo(ProjectCapability.CSharpOrVisualBasicLanguageService)]
+    internal class RenamerProjectTreeActionHandler : ProjectTreeActionHandlerBase, IFileRenameHandler
+    {
+        private readonly IEnvironmentOptions _environmentOptions;
+        private readonly IUnconfiguredProjectVsServices _projectVsServices;
+        private readonly IProjectThreadingService _threadingService;
+        private readonly UnconfiguredProject _unconfiguredProject;
+        private readonly IVsUIService<IVsExtensibility, IVsExtensibility3> _extensibility;
+        private readonly IVsOnlineServices _vsOnlineServices;
+        private readonly IUserNotificationServices _userNotificationServices;
+        private readonly IWaitIndicator _waitService;
+        private readonly IRoslynServices _roslynServices;
+        private readonly Workspace _workspace;
+        private bool _isLanguageServiceDone;
+        private string _oldFilePath;
+
+        [ImportingConstructor]
+        public RenamerProjectTreeActionHandler(
+            UnconfiguredProject unconfiguredProject,
+            IUnconfiguredProjectVsServices projectVsServices,
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace,
+            IEnvironmentOptions environmentOptions,
+            IUserNotificationServices userNotificationServices,
+            IRoslynServices roslynServices,
+            IWaitIndicator waitService,
+            IVsOnlineServices vsOnlineServices,
+            IProjectThreadingService threadingService,
+            IVsUIService<IVsExtensibility, IVsExtensibility3> extensibility)
+        {
+            _unconfiguredProject = unconfiguredProject;
+            _projectVsServices = projectVsServices;
+            _workspace = workspace;
+            _environmentOptions = environmentOptions;
+            _userNotificationServices = userNotificationServices;
+            _roslynServices = roslynServices;
+            _waitService = waitService;
+            _vsOnlineServices = vsOnlineServices;
+            _threadingService = threadingService;
+            _extensibility = extensibility;
+        }
+
+        public void HandleRename(string oldFilePath, string newFilePath)
+        {
+            if (_oldFilePath == oldFilePath)
+            {
+                _isLanguageServiceDone = true;
+            }
+        }
```

Comment:
Successive renames will throw this logic and overlap.

Reply:
Let me try to understand.
1) 'fileA' -> 'fileB' with slow evaluation
2) 'fileA' -> 'fileC' with fast evaluation
3) evaluation of 'fileC' is completed and HandleRename() is executed
4) _isLanguageServiceDone to true (at this point evaluation for 'fileB' still in progress). 
5) continue the RunAndForget

It seems that we need to check 'oldFilePath' and 'newFilePath'.
If multiple renames, should apply only the latest rename (not the fastest evaluation) ?

Reply:
@davkean I added _languageServiceDone to contain the list of semaphores foreach rename.